### PR TITLE
rbd-nbd: add ability to quiesce on signal

### PR DESCRIPTION
**Pull request by pkalever**
Problem:
-------
Scenario possible most commonly with the new attach command.

Say we had mapped the image and mounted it, and now the application is performing IO to the device, while the IO is ongoing it might be possible that the rbd-nbd process received a detach request or a SIGTERM (this is very common if the rbd-nbd is running inside a container, if there is a container restart then it will result in SIGTERM). It is very much possible that we hit a filesystem corruption because it might be possible that some of the writes are still waiting at the filesystem or the block layers, which have not reached rbd-nbd layer yet.

This can be easily handled:
1. If the application can send sync before killing the rbd-nbd process, but that cannot be guaranteed all the time, because we should be ready for ungraceful shutdown nature.
2. Also we might not want to afford synchronous I/O (mount -osync)

Also what about the IO which might be still continuing on the mount while process is down? We cannot simply assume that the applications are always aware of rbd-nbd process state.

One thing we all agree with is that we should always give highest priority to the data consistency.

Solution:
--------
Freeze the filesystem as part of the detach (SIGTERM handling) so that:
1. All the data is flushed from filesystem till backend, so that data is guaranteed to be consistent across.
2. IO on the mount, while rbd-nbd process is down is blocked waiting for the filesystem.
3. Snapshots on the image, while rbd-nbd process is down can also be consistent.
4. Just in case if the rbd-nbd process cannot be reattached, no problem! we have a safe copy.

By the way, this is provided just as an option `--quiesce-onsignal` (read as quiesce on signal) so that default is still the same as old.

Fixes: https://tracker.ceph.com/issues/51820
Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by pkalever** [comment]
cc: @nixpanic @lxbsz

**Comment by nixpanic** [comment]
@pkalever, great to have this! I would like to know your reasons for making the new quiesce functions optional, and not default?

**Comment by mykaul** [comment]
I hope the hook is similar to the hook that qemu-guest-agent uses - see https://github.com/qemu/qemu/tree/master/scripts/qemu-guest-agent/fsfreeze-hook.d

**Comment by pkalever** [comment]
> I hope the hook is similar to the hook that qemu-guest-agent uses - see https://github.com/qemu/qemu/tree/master/scripts/qemu-guest-agent/fsfreeze-hook.d

Yes, the idea is similar, we run fsfreeze as part of the quiesce hook. Here is the one that goes as part of the package: https://github.com/ceph/ceph/blob/master/src/tools/rbd_nbd/rbd-nbd_quiesce

File: src/tools/rbd_nbd/rbd-nbd.cc [hunk 1]
```diff
@@ -540,6 +572,24 @@ class NBDServer
     }
   }
 
+  void quiesce_onsignal_entry()
+  {
+    while (wait_signal()) {
+      ceph_assert(signal_ctx == true);
```
Comment (lxbsz) [inline src/tools/rbd_nbd/rbd-nbd.cc:578]: Should it be 

```
if (!signal_ctx) {
    break;
}
```

Or in `wait_signal()` will return true only when `signal_ctx` is true ?

**Comment by mykaul** [comment]
> > I hope the hook is similar to the hook that qemu-guest-agent uses - see https://github.com/qemu/qemu/tree/master/scripts/qemu-guest-agent/fsfreeze-hook.d
> 
> Yes, the idea is similar, we run fsfreeze as part of the quiesce hook. Here is the one that goes as part of the package: https://github.com/ceph/ceph/blob/master/src/tools/rbd_nbd/rbd-nbd_quiesce

If it's an extendable hook - that's great. Note that there's a KEP somewhere that discuss the general ability of providing a hook for application-level consistency before taking a snapshot for containers. I hope we can align with it.

**Comment by mykaul** [comment]
> > > I hope the hook is similar to the hook that qemu-guest-agent uses - see https://github.com/qemu/qemu/tree/master/scripts/qemu-guest-agent/fsfreeze-hook.d
> > 
> > 
> > Yes, the idea is similar, we run fsfreeze as part of the quiesce hook. Here is the one that goes as part of the package: https://github.com/ceph/ceph/blob/master/src/tools/rbd_nbd/rbd-nbd_quiesce
> 
> If it's an extendable hook - that's great. Note that there's a KEP somewhere that discuss the general ability of providing a hook for application-level consistency before taking a snapshot for containers. I hope we can align with it.

https://github.com/kubernetes/enhancements/pull/1995

**Comment by pkalever** [comment]
> If it's an extendable hook - that's great. Note that there's a KEP somewhere that discuss the general ability of providing a hook for application-level consistency before taking a snapshot for containers. I hope we can align with it.

Yes, we are aware of it Yaniv, thanks!
The `--quiesce` is optional, that can be disabled for snapshots once we the feature from K8s.

File: src/tools/rbd_nbd/rbd-nbd.cc [hunk 2]
```diff
@@ -1422,10 +1478,16 @@ static void handle_signal(int signum)
...
-  dout(20) << __func__ << ": " << "notifying terminate" << dendl;
-
-  ceph_assert(terminate_event_sock.is_valid());
-  terminate_event_sock.notify();
+  if (!quiesce_onsignal) {
+    dout(20) << __func__ << ": " << "notifying terminate" << dendl;
+    ceph_assert(terminate_event_sock.is_valid());
+    terminate_event_sock.notify();
+  } else {
+    std::lock_guard g_l{g_lock};
```
[the remark this answers is not in the document]
Comment (pkalever) [reply src/tools/rbd_nbd/rbd-nbd.cc:1486]: Sure, I can make a cleanup patch to remove the current use of lock_guard in rbd-nbd.cc to replace it with scoped_lock if that is useful and apply this patch aligning to use of scoped_lock.

File: src/tools/rbd_nbd/rbd-nbd.cc (hunk 1 above)
Comment (pkalever) [reply src/tools/rbd_nbd/rbd-nbd.cc:578]: whether it is signal_ctx or not, once the wait_signal() is unblocked the quiesce_onsignal_entry() has no job to continue after performing the run_quiesce_hook() and run_quiesce_hook(), unless I'm missing something.

Hence breaking unconditionally. Please let me know if I missed anything.

Comment (pkalever) [reply src/tools/rbd_nbd/rbd-nbd.cc:578]: I was thinking you are referring to the `break` at the end of the loop, I get you now. Thanks!

**Comment by pkalever** [comment]
@idryomov @trociny could you please help review this? thanks!

**Comment by trociny** [comment]
@pkalever I am not sure I like the idea very much.

A quiesce hook on a snapshot create event makes sense because the snapshot create may be executed on any host that has access to the cluster, so it may be troublesome to run the hook manually to prepare the VM before snapshotting. There is also another case when snapshot create quiesce hook is useful: when exclusive lock feature is not used and the snapshot create request is not proxied to the writer. In this case the writer will not be aware about new snapshot until its watcher gets the message to refresh the image metadata and in this case it may issue some io with not updated snapshot sequence context, i.e writing to the snapshot. The hook may block io writes for this case.

In the reattach case one will always do this on the host where the image is mapped, and the tool that does reattach is usually the same that does initial map, so technically there is no much difference if you run:
```
rbd map --hook /path/to/hook ...
...
rbd detach
rbd attach
```
or
```
rbd map ...
...
/path/to/hook quiesce
rbd detach
/path/to/hook unquiesce
rbd attach
```
Adding the hook seems to provide some convenience for users, but it might also lead to some issues that are hard to troubleshoot. E.g, you seem to have to run unquiesce on attach unconditionally, even if the user does not want it and does not expect. It might be not a problem with the default hook script (except some weird error messages) but what if the user has a custom script?

Saying this, I will not object to have this merged if other people find this useful and vote for this. I would change the option name `--quiesce-onsignal` in this case though. To me the option suggests that it would be called on any signal, e.g. SIGHUP, not only on a termination signal. Also, users might not know that `rbd detach` is sending the signal, so the hook will be called. I would call  it `--quiesce-on-exit`. And I would extent the hook to accept and check the second argument: the first would be `quiesce|unquiesce` (as it is already now) and the second would be `snapshot|exit|attach` (or may be `snapshot|reattach`), so the hook script could distinguish the snapshot create and reattach cases.

**Comment by pkalever** [comment]
Hello @trociny 

> @pkalever I am not sure I like the idea very much.

You probably missed looking at this from the container lens :-)

> 
> A quiesce hook on a snapshot create event makes sense because the snapshot create may be executed on any host that has access to the cluster, so it may be troublesome to run the hook manually to prepare the VM before snapshotting. There is also another case when snapshot create quiesce hook is useful: when exclusive lock feature is not used and the snapshot create request is not proxied to the writer. In this case the writer will not be aware about new snapshot until its watcher gets the message to refresh the image metadata and in this case it may issue some io with not updated snapshot sequence context, i.e writing to the snapshot. The hook may block io writes for this case.

Yes thanks for the history and context, it is certainly interesting and helpful.

> 
> In the reattach case one will always do this on the host where the image is mapped, and the tool that does reattach is usually the same that does initial map, so technically there is no much difference if you run:
> 
> ```
> rbd map --hook /path/to/hook ...
> ...
> rbd detach
> rbd attach
> ```
> 
> or
> 
> ```
> rbd map ...
> ...
> /path/to/hook quiesce
> rbd detach
> /path/to/hook unquiesce
> rbd attach
> ```

Although the rbd-nbd process and the application mount resides on the same node in the ceph-csi use case they both run in two different pods. Also, no one in the stack will be able to predict when an ungrateful/unplanned nodeplugin pod shutdown running rbd-nbd will happen, hence no one is in a better place than rbd-nbd to trigger the hook. Since the rbd-nbd receives the SIGTERM signal, as part of rbd-nbd's signal handler it can do this magic!

> 
> Adding the hook seems to provide some convenience for users, but it might also lead to some issues that are hard to troubleshoot. E.g, you seem to have to run unquiesce on attach unconditionally, even if the user does not want it and does not expect. It might be not a problem with the default hook script (except some weird error messages) but what if the user has a custom script?

I didn't find a way to check `fsfreeze` status, but I would have improved the hook script to accommodate a defence check in the `unquiesce` to see if the mount is already frozen or not. I would expect the custom scripts to come with such defence checks.

> 
> Saying this, I will not object to have this merged if other people find this useful and vote for this. I would change the option name `--quiesce-onsignal` in this case though. To me the option suggests that it would be called on any signal, e.g. SIGHUP, not only on a termination signal. Also, users might not know that `rbd detach` is sending the signal, so the hook will be called. I would call it `--quiesce-on-exit`. And I would extent the hook to accept and check the second argument: the first would be `quiesce|unquiesce` (as it is already now) and the second would be `snapshot|exit|attach` (or may be `snapshot|reattach`), so the hook script could distinguish the snapshot create and reattach cases.

Thank you! This would be really helpful for ceph-csi use case.

I'm not really committed to the option naming used in the patch here, hence I would welcome any better suggestions.
`--quiesce-on-exit` looks good, but sounds like something that also runs as part of `unmap`? (give it a thought)

Yes! providing more options to segregate hook logic adds flexibility. For the use case I was addressing, it needs fsfreeze for both snapshot and signal handler, hence I didn't look much into it. But certainly, we can have more cases for the hook script.

That all said,
If the use case that I described above convinces you, please help with the code review.

Thanks!

**Comment by trociny** [comment]
> I'm not really committed to the option naming used in the patch here, hence I would welcome any better suggestions.
> `--quiesce-on-exit` looks good, but sounds like something that also runs as part of `unmap`? (give it a thought)

Then may be `--quiesce-on-detach`?

> If the use case that I described above convinces you, please help with the code review.

I am not sure it convinced me but I will not object and will review if other people are fine with it.

**Comment by pkalever** [comment]
Lots of discussion about this happened at https://github.com/ceph/ceph-csi/pull/2394

Yet another case:

To address the reattach gracefully, we were suggested to live with `io-timeout=0` by default, but also as @idryomov rightly mentioned at https://github.com/ceph/ceph-csi/pull/2394#issuecomment-903636068 

> Kernel 4.19.182 doesn't handle io-timeout=0 correctly because it lacks torvalds/linux@2da22da. In general, the nbd driver in older kernels is quite broken so you may need to constrain nbd tests to e.g. CentOS 8.4.

I'm hoping for kernels without torvalds/linux@2da22da we should use the quiesce on signal mechanism https://github.com/ceph/ceph/pull/42609


cc: @nixpanic @Madhu-1

**Comment by idryomov** [comment]
Apart from the "why" questions in https://github.com/ceph/ceph-csi/pull/2394#issuecomment-904615183, one major issue I have with the "how" is that freezing the filesystem means putting the user workload processes into D (uninterruptible sleep) state.  Until something comes back and unfreezes there is no way to get rid of them (other than to reboot the node).  Unlike with snapshot creation where unfreezing is handled by the same daemon and the only thing that can potentially prevent it is a daemon crash, here it is really vague.  IMO this is even worse than the pre-`io-timeout=0` I/O errors -- in that case the user at least got a clear error and none of the unkillable processes business.

**Comment by Madhu-1** [comment]
@idryomov hope Prasanna already answered all the questions, what I came to know is that without this fix we cannot support rbd-nbd on all the kernel versions in cephcsi. @pkalever is my understanding is correct?  @idryomov How we can take this forward.

**Comment by pkalever** [comment]
>  @pkalever is my understanding is correct?

@Madhu-1 absolutely!

**Pull request closed without merging**